lib/datadog/appsec/contrib/graphql/appsec_trace.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+require 'json'
+require_relative 'ext'
+require_relative 'gateway/multiplex'
+require_relative '../../instrumentation/gateway'
+
+module Datadog
+  module AppSec
+    module Contrib
+      module GraphQL
+        # These methods will be called by the GraphQL runtime to send the variables to the WAF.
+        # We actually don't need to create any span/trace.
+        module AppSecTrace
+          def execute_multiplex(multiplex:)
+            return super unless Datadog::AppSec.enabled?
+
+            gateway_multiplex = Gateway::Multiplex.new(multiplex)
+
+            multiplex_return, multiplex_response = Instrumentation.gateway.push('graphql.multiplex', gateway_multiplex) do
+              super
+            end
+
+            # Returns an error * the number of queries so that the entire multiplex is blocked
+            if multiplex_response
+              blocked_event = multiplex_response.find { |action, _options| action == :block }
+              if blocked_event
+                multiplex_return = []
+                gateway_multiplex.queries.each do |query|
+                  query_result = ::GraphQL::Query::Result.new(
+                    query: query,
+                    values: JSON.parse(AppSec::Response.content_json)
+                  )
+                  multiplex_return << query_result
```

Comment:
We could extract it but I believe it would become un-necessarily more complex. `AppSec::Response.negotiate` was made for HTTP-level frameworks, selecting the type of response the client configured (JSON, HTML or Plain text). But GraphQL only return JSON, so we'd have to make a special case to always enforce JSON when it is GraphQL, and also duplicate the resulting JSON times the number of queries in the multiplex.

Reply:
```ruby
multiplex_return = appsec_response(some, args)
```

and

```ruby
private

def appsec_response(some, args)
  gateway_multiplex.queries.map do |query|
    query_result = ::GraphQL::Query::Result.new(
      query: query,
      values: JSON.parse(AppSec::Response.content_json)
    )
  end
end
```